Fix #2569 (partial): Add guest interactions with collection player

Ian91 opened the pull request:
This PR fixes part 2 of #2569. 

* See comment on line 107 of collections.js and on line 56 of collectionPlayer.js. Can the assignee of part 3 use this partial test (and partial function), or should I delete them and let them start clean on part 3?

* The order of nodes in welcome_to_collections.yaml was changed per discussion with @seanlip.

Commit:
Fix #2569 (partial): Add guest interactions with collection player (part 2 of issue)

seanlip commented on core/tests/protractor/collections.js:
```diff
@@ -30,7 +31,6 @@ describe('Collections', function() {
       '.protractor-test-reload-collection-button')).first().click();
     general.acceptAlert();
     browser.waitForAngular();
-    admin.reloadAllExplorations();
```
Any reason to delete this?

Ian91 replied:
It looks like this (and the similar problem immediately below) is some kind of conflict with PR #3507. The code that appears to be deleted was not present before my commit. I followed the instructions at "Contributing code to Oppia", but I did notice my branch showed something like "1 commit ahead and 1 commit behind develop". Is there a name for this problem (a type of merge conflict) that I could look up to understand better?

seanlip replied:
That's a bit odd -- I'd have thought that GitHub would handle the merge automatically. Can you try merging upstream/develop into your branch and pushing again?

seanlip commented on core/tests/protractor/collections.js:
```diff
@@ -55,10 +57,6 @@ describe('Collections', function() {
     collectionEditor.addExistingExploration('0');
     collectionEditor.addExistingExploration('4');
     collectionEditor.addExistingExploration('13');
-    // Search and add existing explorations.
```
Why delete this?

seanlip commented on core/tests/protractor_utils/collectionPlayer.js:
```diff
@@ -0,0 +1,106 @@
+  player.submitAnswer('NumericInput', 121);
+  browser.waitForAngular();
+  player.clickThroughToNextCard();
+  // Function needs to be finished
```
Oh, you're going to run into trouble here. The next interaction is a map and it's going to be quite hard to test that.

I'm starting to wonder whether it might be better to create a dummy collection with two or three trivial explorations instead (maybe 1 text input state and then it ends), and then play through that. The explorations in the demo collection is too complex for a test like this, and that complexity doesn't actually relate directly to what is being tested here. In the meantime, perhaps you could just delete this function since you are not using it?

seanlip commented on core/tests/protractor/collections.js:
```diff
@@ -83,6 +81,32 @@ describe('Collections', function() {
+    collectionPlayer.verifyProfileDropdown();
+    // Verify that there are 5 uncompleted explorations (initial state).
+    collectionPlayer.verifyNumUncompletedExplorations(5);
+    // This test (i.e. part 3) needs to be finished
```
Format should be:

    // TODO(Ian91): ...

and the "..." should specify, in detail, exactly what needs to be done, so that someone with almost no context can pick it up and do it.

Ian91 commented:
Oops, looks like I did something strange with my branches and probably lost the CSS classes that I had added (which I assume will cause the Travis build to fail). I won't be able to fix this until Thursday, but the two HTML files that I need to add in the next commit (with the added classes) can be seen [here](https://github.com/Ian91/oppia/commit/d721358bc1dc615e9bc681a485d005c32841b6a7) (but ignore the change on lines 379 and 380 of collection_player.html).

seanlip reviewed (changes requested):
Hi @Ian91, thanks! I'll do another pass on this tomorrow when I'm more awake, but in the meantime here are a few comments to address.

seanlip commented on core/tests/protractor/collections.js:
```diff
@@ -39,6 +39,8 @@ describe('Collections', function() {
     );
     users.logout();
   });
+ 
```
Nit: please remove these extra newlines.

seanlip commented on core/tests/protractor/collections.js:
```diff
@@ -83,6 +81,32 @@ describe('Collections', function() {
+  it('interacts with the collection player as a guest', function() {
+    collectionPlayer.goToDemoCollection();
+    // Check that we are in the guest state (sign in button visible).
+    collectionPlayer.verifySignInButton();
```
This doesn't seem particularly collectionPlayer-specific. Perhaps it should go in `general`? And maybe we can call it verifyUserIsLoggedOut().

seanlip commented on core/tests/protractor/collections.js:
```diff
@@ -83,6 +81,32 @@ describe('Collections', function() {
+    // Verify that there are 5 uncompleted explorations (initial state).
+    collectionPlayer.verifyNumUncompletedExplorations(5);
+    // Play the third exploration in the demo collection.
+    collectionPlayer.playDemoExplorationThree();
```
The problem with this is that playing a specific exploration isn't something that's really specific to the collection player. I'd put the actions to go through the exploration here (in the test itself) -- see also my other comment about creating your own collection whose explorations are "simple" and easily completeable.

seanlip commented on core/tests/protractor/collections.js:
```diff
@@ -83,6 +81,32 @@ describe('Collections', function() {
+    // Play the third exploration in the demo collection.
+    collectionPlayer.playDemoExplorationThree();
+    // Verify that the suggested exploration is the next in the collection.
+    collectionPlayer.verifySuggestedExplorationAsGuest();
```
collectionPlayer.expectRecommendedExplorationToBe(explorationTitle)

seanlip commented on core/tests/protractor/collections.js:
```diff
@@ -83,6 +81,32 @@ describe('Collections', function() {
   });
 
+  it('interacts with the collection player as a guest', function() {
+    collectionPlayer.goToDemoCollection();
```
collectionPlayer.goToCollection(collectionId)

seanlip commented on core/tests/protractor/collections.js:
```diff
@@ -83,6 +81,32 @@ describe('Collections', function() {
+    users.login('alice@collections.com');
+    collectionPlayer.goToDemoCollection();
+    // Verify that profile dropdown thumbnail appears, since we are logged in.
+    collectionPlayer.verifyProfileDropdown();
```
general.verifyUserIsLoggedIn()

seanlip commented on core/tests/protractor_utils/collectionPlayer.js:
```diff
@@ -0,0 +1,106 @@
+
+// Verify that the suggested exploration is the expected one.
+var verifySuggestedExplorationAsGuest = function() {
+  explTitle = element.all(by.css(
```
Here and below: missing `var`?

seanlip commented on core/tests/protractor_utils/collectionPlayer.js:
```diff
@@ -0,0 +1,106 @@
+// Verify that the "SIGN IN" button appears correctly in guest's view.
+var verifySignInButton = function() {
+  var text = element(by.css('.protractor-test-sign-in-button')).getText();
+  expect(text).toBe('SIGN IN');
```
Perhaps also add:

  expect(element.all(by.css(
    '.protractor-test-profile-dropdown')).count()).toEqual(0);

seanlip reviewed (changes requested):
OK, I took a full pass. The basic commands look good, but I have a few conceptual comments about the new collection player e2e test utility API. Please let me know if you have any questions!

anmolshkl commented:
@Ian91 ping?

Ian91 commented:
@seanlip @anmolshkl Thanks for checking in. I can push a commit addressing the comments by Sunday 10 PM PST, and I'll make sure to stay in touch with @seanlip for any questions. I understand if it needs to be reassigned for faster completion, but if not, I'll make sure to stay in touch daily as I continue working on it.

seanlip commented:
Sounds good, thansk @Ian91!

seanlip commented:
Hm. @Ian91, are you still working on this? I think Sunday's come and gone...

jacobdavis11 commented:
I'm closing this PR. Please open a new one whenever you'd like to resume.
